jax/_src/pallas/mosaic/interpret.py

```diff
@@ -0,0 +1,694 @@
+
+  # TODO(jburnim): Check/fix so that IDs for barrier semaphores do not conflict
+  # with IDs for regular or DMA semaphores.  (For example, store them in a
+  # different table.)
+  with shared_memory.lock:
+    semaphore_id = collective_id
+    if not semaphore_id in shared_memory.sem:
+      shared_memory.sem[semaphore_id] = Semaphore()
+
+  return np.int16(semaphore_id)
+
+def _transform_slice_or_index(slice_or_idx):
+  if isinstance(slice_or_idx, int):
+    return slice_or_idx
+  else:
+    start, size, stride = (
+        slice_or_idx.start, slice_or_idx.size, slice_or_idx.stride)
+    return slice(start, start + size * stride, stride)
+
+def transform_array(x, transforms):
+  for transform in transforms:
+    # For now, assume only NDIndexer transforms.
+    x = x[tuple(_transform_slice_or_index(i) for i in transform.indices)]
+  return x
+
+def get(device_id, memory_space, buffer_id, transforms):
+  # NOTE: buffer_id is a pair of an int ID and a tuple of transforms
+  # TODO(jburnim): Clean this up.
+  buffer_id, transforms = buffer_id[0], buffer_id[1] + transforms
+
+  device_id = tuple(int(x) for x in device_id)
+  memory_space = TPU_MEMORY_SPACE_NAMES[int(memory_space)]
+  buffer_id = int(buffer_id)
+  transforms = jax.tree.map(int, transforms)
+  with shared_memory.lock:
+    return transform_array(
+        shared_memory.mem[(memory_space, buffer_id, device_id)], transforms
+    ).copy()
+
+def store(device_id, memory_space, buffer_id, transforms, val):
+  # NOTE: buffer_id is a pair of an int ID and a tuple of transforms.
+  buffer_id, transforms = buffer_id[0], buffer_id[1] + transforms
+
+  device_id = tuple(int(x) for x in device_id)
+  memory_space = TPU_MEMORY_SPACE_NAMES[int(memory_space)]
+  buffer_id = int(buffer_id)
+  val = np.asarray(val)
+  with shared_memory.lock:
+    if transforms:
+      transform_array(
+          shared_memory.mem[(memory_space, buffer_id, device_id)],
+          transforms
+      )[:] = val
+    else:
+      shared_memory.mem[(memory_space, buffer_id, device_id)] = val
+
+def swap(device_id, memory_space, buffer_id, transforms, val):
+  # NOTE: buffer_id is a pair of an int ID and a tuple of transforms.
+  buffer_id, transforms = buffer_id[0], buffer_id[1] + transforms
+
+  device_id = tuple(int(x) for x in device_id)
+  memory_space = TPU_MEMORY_SPACE_NAMES[int(memory_space)]
+  buffer_id = int(buffer_id)
+  transforms = jax.tree.map(int, transforms)
+  val = np.array(val)
+
+  with shared_memory.lock:
+    mem_val = shared_memory.mem[(memory_space, buffer_id, device_id)].copy()
+  # TODO(jburnim): Do this transform with NumPy arrays?
+  result, result_val = state_discharge.transform_swap_array(
+      jnp.array(mem_val), transforms, jnp.array(val))
+  result_val = np.array(result_val)
+  with shared_memory.lock:
+    shared_memory.mem[(memory_space, buffer_id, device_id)] = result_val
+
+  return np.array(result)
+
+def execute_dma(src, dst, send_sem, recv_sem):
```

Comment:
Add type annotations for src/dst?

Reply:
I added a comment for now.

I think this function may go away when the code is updated to execute DMAs asynchronously.